File: 1448-CountGoodNodesInBinaryTree/CountGoodNodesInBinaryTree.py
# Python 3
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        count  = 1
        
        def dfs(node, maxVal):
            
            nonlocal count
            
            if node:
            
                if node.val >= maxVal:
                    count += 1
            
                if node.left:
                    dfs(node.left, max(node.val, maxVal))
                
                if node.right:
                    dfs(node.right, max(node.val, maxVal))
                
        dfs(root.left, root.val)
        dfs(root.right, root.val)
        
        return count

File: 1448-CountGoodNodesInBinaryTree/CountGoodNodesInBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // Reduce memory usage
    // Might be bad for multi-threading?
    int max = -10000;
    int count = 0;
    
    int maxNodes(TreeNode* root , int mx){
        if (root){
            if (root->val >= mx){
                return 1 + maxNodes(root->left, root->val)+maxNodes(root->right, root->val);
            } else {
                return maxNodes(root->left, mx)+maxNodes(root->right, mx);
            }
        }
        return 0;
        
    }
    
    int goodNodes(TreeNode* root) {
        if (root){
            count += maxNodes(root, INT_MIN);
        }
        return count;
    }
    
    
};
